From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: way-zer <himc.wicp@gmail.com>
Date: Thu, 25 Apr 2024 14:49:41 +0800
Subject: [PATCH] =?UTF-8?q?OC:=20=E4=BC=98=E5=8C=96Batch=E6=80=A7=E8=83=BD?=
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

大幅优化渲染性能
UPDATE1: 增加配置项，默认关闭
UPDATE2: 优化性能；我真的已经尽力了，剩下的只能怪java慢了
UPDATE3: 完全重写了sort，又提升了一倍sort性能
UPDATE4: 修复一些bug，并优化性能。这次重写了IntIntMap。(排序算法相当于完全重写完了)
UPDATE5: 修复文字渲染错误

sorted不写回，减少开销。
way-zer <himc.wicp@gmail.com> on 2024/5/28

新功能：顶点合并
way-zer <himc.wicp@gmail.com> on 2024/7/23
---
 .../arc/graphics/g2d/MySortedSpriteBatch.java | 237 ++++++-
 .../arc/graphics/g2d/SortedSpriteBatch2.java  | 668 ++++++++++++++++++
 core/src/mindustry/graphics/MultiPacker.java  |   2 +-
 core/src/mindustryX/features/RenderExt.java   |   3 +
 4 files changed, 907 insertions(+), 3 deletions(-)
 create mode 100644 core/src/arc/graphics/g2d/SortedSpriteBatch2.java

diff --git a/core/src/arc/graphics/g2d/MySortedSpriteBatch.java b/core/src/arc/graphics/g2d/MySortedSpriteBatch.java
index feca84a8eec0f7d514f9aa6be6f82134d8622721..9f56c24264dcb7d0ccc5db464d6d6deaa1d08f20 100644
--- a/core/src/arc/graphics/g2d/MySortedSpriteBatch.java
+++ b/core/src/arc/graphics/g2d/MySortedSpriteBatch.java
@@ -1,10 +1,29 @@
 package arc.graphics.g2d;
 
 import arc.graphics.*;
+import arc.math.*;
 import mindustryX.features.*;
 
+import java.util.*;
+
 //MDTX: add some DebugUtil count.
-public class MySortedSpriteBatch extends SortedSpriteBatch{
+//MDTX(WayZer): renderSort
+public class MySortedSpriteBatch extends SortedSpriteBatch2{
+    private static final boolean DEBUG = false;
+    private static final int PRIME1 = 0xbe1f14b1;
+    private static final int PRIME2 = 0xb4b82e39;
+    int[] extraZ = new int[InitialSize];
+    //增加小的delta，来保持原来的前后顺序
+    int orderZ = 0;
+    int hashZ = 0;//打乱hash值，来检查渲染异常
+
+    @Override
+    protected void z(float z){
+        if(this.z == z) return;
+        orderZ = 0;
+        super.z(z);
+    }
+
     @Override
     protected void flush(){
         DebugUtil.lastFlushCount++;
@@ -13,7 +32,6 @@ public class MySortedSpriteBatch extends SortedSpriteBatch{
 
     @Override
     protected void flushRequests(){
-        if(flushing) return;
         DebugUtil.lastVertices += numVertices / 6;
         DebugUtil.lastDrawRequests += numRequests;
         super.flushRequests();
@@ -24,4 +42,219 @@ public class MySortedSpriteBatch extends SortedSpriteBatch{
         DebugUtil.lastSwitchTexture++;
         super.switchTexture(texture);
     }
+
+    @Override
+    protected void expandRequests(){
+        super.expandRequests();
+        extraZ = Arrays.copyOf(extraZ, requestZ.length);
+    }
+
+    @Override
+    protected void draw(Texture texture, float[] spriteVertices, int offset, int count){
+        super.draw(texture, spriteVertices, offset, count);
+        if(sort && !flushing && RenderExt.renderSort){
+            int h = texture.hashCode();
+            if(DEBUG){
+                h = (h + hashZ) * PRIME1;
+                h = h ^ (h >>> 16);
+            }
+            extraZ[numRequests - 1] = ((orderZ++) << 16) | (h & 0xffff);
+        }
+    }
+
+    @Override
+    protected void draw(Runnable request){
+        super.draw(request);
+        if(sort && !flushing && RenderExt.renderSort){
+            int h = DEBUG ? hashZ : 0;
+            extraZ[numRequests - 1] = ((orderZ++) << 16) | (h & 0xffff);
+        }
+    }
+
+    @Override
+    protected void sortRequests(){
+        if(!RenderExt.renderSort){
+            super.sortRequests();
+            return;
+        }
+        hashZ = DEBUG ? Float.floatToIntBits((float)Math.random()) : 0;
+
+        int numRequests = this.numRequests;
+        int[] arr = this.requestZ, extraZ = this.extraZ;
+        sortMap(arr, numRequests);
+        sortMap(extraZ, numRequests);
+        for(int i = 0; i < numRequests; i++){
+            arr[i] = (arr[i] << 16) | extraZ[i];
+        }
+        countingSortMap(arr, numRequests);//arr is loc now;
+
+        if(copy.length < requests.length) copy = new DrawRequest[requests.length];
+        final DrawRequest[] items = requests, dest = copy;
+        for(int i = 0; i < numRequests; i++){
+            dest[arr[i]] = items[i];
+        }
+    }
+
+    private static final IntIntMap vMap = new IntIntMap(10000, 0.25f);
+    private static int[] order = new int[1000];
+    private static int[] order2 = new int[1000];
+
+    /**
+     * 将输入arr重映射到有序的[0,unique)域
+     * @param arr 待排序数组，输出会映射为id值
+     * @return unique
+     */
+    private static int sortMap(int[] arr, int len){
+        var map = MySortedSpriteBatch.vMap;
+        int[] order = MySortedSpriteBatch.order;
+        map.clear();
+        int unique = 0;
+        for(int i = 0; i < len; i++){
+            int v = arr[i];
+            int id = map.getOrPut(v, unique);
+            arr[i] = id;//arr现在表示id
+            if(id == unique){
+                if(order.length <= unique){
+                    order = Arrays.copyOf(order, unique << 1);
+                }
+                order[unique] = v;
+                unique++;
+            }
+        }
+        MySortedSpriteBatch.order = order;
+
+        //对z值排序
+        Arrays.sort(order, 0, unique);//order -> z
+
+        //arr中储存order
+        int[] order2 = MySortedSpriteBatch.order2;//id -> order
+        if(order2.length < order.length){
+            order2 = new int[order.length];
+            MySortedSpriteBatch.order2 = order2;
+        }
+        for(int i = 0; i < unique; i++){
+            order2[map.getOrPut(order[i], -1)] = i;
+        }
+        for(int i = 0; i < len; i++){
+            arr[i] = order2[arr[i]];
+        }
+        return unique;
+    }
+
+    /**
+     * 计数排序
+     * @param arr 待排序数组，输出为新loc
+     */
+    private static void countingSortMap(int[] arr, int len){
+        int[] order = MySortedSpriteBatch.order, counts = MySortedSpriteBatch.order2;
+        var map = MySortedSpriteBatch.vMap;//z->id
+        map.clear();
+        int unique = 0;
+        for(int i = 0; i < len; i++){
+            int v = arr[i];
+            int id = map.getOrPut(v, unique);
+            arr[i] = id;//arr现在表示id
+            if(id == unique){
+                if(unique >= counts.length){
+                    order = Arrays.copyOf(order, unique << 1);
+                    counts = Arrays.copyOf(counts, unique << 1);
+                    MySortedSpriteBatch.order = order;
+                    MySortedSpriteBatch.order2 = counts;
+                }
+                order[unique] = v;
+                counts[unique] = 1;
+                unique++;
+            }else counts[id]++;
+        }
+
+        //对z值排序
+        Arrays.sort(order, 0, unique);//order -> z
+
+        //将counts转换为locs(每个id起始位置)
+        for(int i = 0, loc = 0; i < unique; i++){
+            int id = map.getOrPut(order[i], -1);
+            int c = counts[id];
+            counts[id] = loc;
+            loc += c;
+        }
+        //arr现在表示新目的地
+        for(int i = 0; i < len; i++){
+            arr[i] = counts[arr[i]]++;
+        }
+    }
+
+    static public class IntIntMap{
+        private int[] keys;
+        private boolean hasZero;
+        private int[] values;
+        private int zeroValue;
+        private int size; // 哈希表中的元素数量
+
+        private int capacity, maxSize;
+        private float loadFactor;
+        private int mask, hashShift;
+
+        public IntIntMap(int capacity, float loadFactor){
+            setCapacity(capacity, loadFactor);
+        }
+
+        private int hash(int key){
+            key *= PRIME2;
+            return (key ^ key >>> hashShift);
+        }
+
+        public int getOrPut(int key, int defaultValue){
+            if(key == 0){
+                if(hasZero) return zeroValue;
+                zeroValue = defaultValue;
+                hasZero = true;
+                return defaultValue;
+            }
+            int mask = this.mask;
+            int index = hash(key) & mask;
+            int[] keys = this.keys;
+            while(keys[index] != 0){
+                if(keys[index] == key){// 键找到
+                    return values[index];
+                }
+                index = (index + 1) & mask;
+            }
+            //键不存在
+            keys[index] = key;
+            values[index] = defaultValue;
+            size++;
+            if(size > maxSize) setCapacity(capacity << 1, loadFactor);
+            return defaultValue;
+        }
+
+        private void setCapacity(int capacity, float loadFactor){
+            capacity = Mathf.nextPowerOfTwo(capacity);
+            this.capacity = capacity;
+            this.loadFactor = loadFactor;
+            maxSize = (int)(capacity * loadFactor);
+            int mask = this.mask = capacity - 1;
+            hashShift = 31 - Integer.numberOfTrailingZeros(capacity);
+
+            int[] oldKeys = keys, oldValues = values;
+            int[] keys = this.keys = new int[capacity];
+            int[] values = this.values = new int[capacity];
+            if(oldKeys == null || oldValues == null) return;
+            for(int i = 0; i < oldKeys.length; i++){
+                if(oldKeys[i] == 0) continue;
+                int index = hash(oldKeys[i]) & mask;
+                while(keys[index] != 0){
+                    index = (index + 1) & mask;
+                }
+                keys[index] = oldKeys[i];
+                values[index] = oldValues[i]; // 插入或更新值
+            }
+        }
+
+        private void clear(){
+            Arrays.fill(keys, 0);
+            Arrays.fill(values, 0);
+            size = 0;
+            hasZero = false;
+        }
+    }
 }
diff --git a/core/src/arc/graphics/g2d/SortedSpriteBatch2.java b/core/src/arc/graphics/g2d/SortedSpriteBatch2.java
new file mode 100644
index 0000000000000000000000000000000000000000..21cea70ba788c955e9324f7cedca3b385b3a0282
--- /dev/null
+++ b/core/src/arc/graphics/g2d/SortedSpriteBatch2.java
@@ -0,0 +1,668 @@
+package arc.graphics.g2d;
+
+import arc.*;
+import arc.graphics.*;
+import arc.graphics.gl.*;
+import arc.math.*;
+import arc.math.geom.*;
+import arc.struct.*;
+import arc.util.*;
+import mindustryX.features.*;
+
+import java.util.*;
+import java.util.concurrent.*;
+
+//MDTX modified type of requestZ, from float[] to int[]
+//MDTX: sorted requests store in `copy`(instead write back), reduce one copy and prevent memory fragment
+//MDTX(WayZer, 2024/7/19): optimize req.vertices
+
+/** Fast sorting implementation written by zxtej. Don't ask me how it works. */
+public class SortedSpriteBatch2 extends SpriteBatch{
+    public static class DrawRequest{
+        float color, mixColor;
+        int verticesOffset, verticesLength;
+        Texture texture;
+        Blending blending;
+        Runnable run;
+    }
+
+    private static final float[] EXISTED = new float[0];
+    protected static final int InitialSize = 10000;
+    static ForkJoinHolder commonPool;
+    boolean multithreaded = (Core.app.getVersion() >= 21 && !Core.app.isIOS()) || Core.app.isDesktop();
+
+    protected boolean sort, flushing;
+    protected DrawRequest[] requests = new DrawRequest[InitialSize], copy = new DrawRequest[0];
+    protected int[] requestZ = new int[InitialSize];
+    protected float[] requestVertices = new float[SPRITE_SIZE * InitialSize];
+    protected int numRequests = 0, numVertices = 0;
+    int[] contiguous = new int[2048], contiguousCopy = new int[2048];
+    protected int intZ = Float.floatToRawIntBits(z + 16f);
+
+    {
+        for(int i = 0; i < requests.length; i++){
+            requests[i] = new DrawRequest();
+        }
+
+        if(multithreaded){
+            try{
+                commonPool = new ForkJoinHolder();
+            }catch(Throwable t){
+                multithreaded = false;
+            }
+        }
+    }
+
+    @Override
+    protected void setSort(boolean sort){
+        if(this.sort != sort){
+            flush();
+        }
+        this.sort = sort;
+    }
+
+    @Override
+    protected void setShader(Shader shader, boolean apply){
+        if(!flushing && sort){
+            throw new IllegalArgumentException("Shaders cannot be set while sorting is enabled. Set shaders inside Draw.run(...).");
+        }
+        super.setShader(shader, apply);
+    }
+
+    @Override
+    protected void setBlending(Blending blending){
+        this.blending = blending;
+    }
+
+    @Override
+    protected void z(float z){
+        if(this.z == z) return;
+        super.z(z);
+        intZ = Float.floatToRawIntBits(z + 16f);
+    }
+
+    protected int prepareVertices(int toWrite){
+        int idx = numVertices;
+        if(idx + toWrite > requestVertices.length){
+            requestVertices = Arrays.copyOf(requestVertices, requestVertices.length << 1);
+        }
+        numVertices += toWrite;
+        return idx;
+    }
+
+    @Override
+    protected void draw(Texture texture, float[] spriteVertices, int offset, int count){
+        if(sort && !flushing){
+            //MDTX: shared vertices, no slice
+            int num = numRequests;
+            //MDTX: 合批优化
+            if(RenderExt.renderMerge && num > 1){
+                final DrawRequest last = requests[num - 1];
+                if(last.run == null && last.texture == texture && last.blending == blending && requestZ[num - 1] == intZ){
+                    if(spriteVertices != EXISTED){
+                        int idx = prepareVertices(count);
+                        System.arraycopy(spriteVertices, offset, requestVertices, idx, count);
+                    }
+                    last.verticesLength += count;
+                    return;
+                }
+            }
+            if(num >= this.requests.length) expandRequests();
+            final DrawRequest req = requests[num];
+            if(spriteVertices != EXISTED){
+                int idx = req.verticesOffset = prepareVertices(count);
+                System.arraycopy(spriteVertices, offset, requestVertices, idx, count);
+            }else{
+                req.verticesOffset = offset;
+            }
+            req.verticesLength = count;
+            requestZ[num] = intZ;
+            req.texture = texture;
+            req.blending = blending;
+            req.run = null;
+            numRequests++;
+        }else{
+            super.draw(texture, spriteVertices, offset, count);
+        }
+    }
+
+    protected final void draw(TextureRegion region, float x, float y, float originX, float originY, float width, float height, float rotation){
+        if(!sort || flushing){
+            super.draw(region, x, y, originX, originY, width, height, rotation);
+            return;
+        }
+        int idx = prepareVertices(SPRITE_SIZE);
+        region2vertices(requestVertices, idx, region, x, y, originX, originY, width, height, rotation);
+        draw(region.texture, EXISTED, idx, SPRITE_SIZE);
+    }
+
+
+    protected final void region2vertices(float[] vertices, int idx, TextureRegion region, float x, float y, float originX, float originY, float width, float height, float rotation){
+        float u = region.u;
+        float v = region.v2;
+        float u2 = region.u2;
+        float v2 = region.v;
+
+        float color = this.colorPacked;
+        float mixColor = this.mixColorPacked;
+
+        if(!Mathf.zero(rotation)){
+            //bottom left and top right corner points relative to origin
+            float worldOriginX = x + originX;
+            float worldOriginY = y + originY;
+            float fx = -originX;
+            float fy = -originY;
+            float fx2 = width - originX;
+            float fy2 = height - originY;
+
+            // rotate
+            float cos = Mathf.cosDeg(rotation);
+            float sin = Mathf.sinDeg(rotation);
+
+            float x1 = cos * fx - sin * fy + worldOriginX;
+            float y1 = sin * fx + cos * fy + worldOriginY;
+            float x2 = cos * fx - sin * fy2 + worldOriginX;
+            float y2 = sin * fx + cos * fy2 + worldOriginY;
+            float x3 = cos * fx2 - sin * fy2 + worldOriginX;
+            float y3 = sin * fx2 + cos * fy2 + worldOriginY;
+            float x4 = x1 + (x3 - x2);
+            float y4 = y3 - (y2 - y1);
+
+            vertices[idx] = x1;
+            vertices[idx + 1] = y1;
+            vertices[idx + 2] = color;
+            vertices[idx + 3] = u;
+            vertices[idx + 4] = v;
+            vertices[idx + 5] = mixColor;
+
+            vertices[idx + 6] = x2;
+            vertices[idx + 7] = y2;
+            vertices[idx + 8] = color;
+            vertices[idx + 9] = u;
+            vertices[idx + 10] = v2;
+            vertices[idx + 11] = mixColor;
+
+            vertices[idx + 12] = x3;
+            vertices[idx + 13] = y3;
+            vertices[idx + 14] = color;
+            vertices[idx + 15] = u2;
+            vertices[idx + 16] = v2;
+            vertices[idx + 17] = mixColor;
+
+            vertices[idx + 18] = x4;
+            vertices[idx + 19] = y4;
+            vertices[idx + 20] = color;
+            vertices[idx + 21] = u2;
+            vertices[idx + 22] = v;
+            vertices[idx + 23] = mixColor;
+        }else{
+            float fx2 = x + width;
+            float fy2 = y + height;
+
+            vertices[idx] = x;
+            vertices[idx + 1] = y;
+            vertices[idx + 2] = color;
+            vertices[idx + 3] = u;
+            vertices[idx + 4] = v;
+            vertices[idx + 5] = mixColor;
+
+            vertices[idx + 6] = x;
+            vertices[idx + 7] = fy2;
+            vertices[idx + 8] = color;
+            vertices[idx + 9] = u;
+            vertices[idx + 10] = v2;
+            vertices[idx + 11] = mixColor;
+
+            vertices[idx + 12] = fx2;
+            vertices[idx + 13] = fy2;
+            vertices[idx + 14] = color;
+            vertices[idx + 15] = u2;
+            vertices[idx + 16] = v2;
+            vertices[idx + 17] = mixColor;
+
+            vertices[idx + 18] = fx2;
+            vertices[idx + 19] = y;
+            vertices[idx + 20] = color;
+            vertices[idx + 21] = u2;
+            vertices[idx + 22] = v;
+            vertices[idx + 23] = mixColor;
+        }
+    }
+
+    @Override
+    protected void draw(Runnable request){
+        if(sort && !flushing){
+            if(numRequests >= requests.length) expandRequests();
+            final DrawRequest req = requests[numRequests];
+            req.run = request;
+            req.blending = blending;
+            req.mixColor = mixColorPacked;
+            req.color = colorPacked;
+            requestZ[numRequests] = intZ;
+            req.texture = null;
+            numRequests++;
+        }else{
+            super.draw(request);
+        }
+    }
+
+    protected void expandRequests(){
+        final DrawRequest[] requests = this.requests, newRequests = Arrays.copyOf(requests, requests.length * 7 / 4);
+        for(int i = requests.length; i < newRequests.length; i++){
+            newRequests[i] = new DrawRequest();
+        }
+        this.requests = newRequests;
+        this.requestZ = Arrays.copyOf(requestZ, newRequests.length);
+    }
+
+    @Override
+    protected void flush(){
+        if(!flushing){
+            flushing = true;
+            flushRequests();
+            flushing = false;
+        }
+        super.flush();
+    }
+
+    protected void flushRequests(){
+        if(numRequests == 0) return;
+        sortRequests();
+        float preColor = colorPacked, preMixColor = mixColorPacked;
+        Blending preBlending = blending;
+
+        float[] vertices = this.requestVertices;
+        DrawRequest[] r = copy;//MDTX: 'copy' instead requests
+        int num = numRequests;
+        for(int j = 0; j < num; j++){
+            final DrawRequest req = r[j];
+
+            colorPacked = req.color;
+            mixColorPacked = req.mixColor;
+
+            super.setBlending(req.blending);
+
+            if(req.run != null){
+                req.run.run();
+                req.run = null;
+            }else if(req.texture != null){
+                super.draw(req.texture, vertices, req.verticesOffset, req.verticesLength);
+            }else{
+                throw new IllegalStateException("No Region Draw");
+//                    super.draw(req.region, req.x, req.y, req.originX, req.originY, req.width, req.height, req.rotation);
+            }
+        }
+
+        colorPacked = preColor;
+        mixColorPacked = preMixColor;
+        color.abgr8888(colorPacked);
+        mixColor.abgr8888(mixColorPacked);
+        blending = preBlending;
+
+        numRequests = numVertices = 0;
+    }
+
+    protected void sortRequests(){
+        if(multithreaded){
+            sortRequestsThreaded();
+        }else{
+            sortRequestsStandard();
+        }
+    }
+
+    protected void sortRequestsThreaded(){
+        final int numRequests = this.numRequests;
+        final int[] itemZ = requestZ;
+
+        int[] contiguous = this.contiguous;
+        int ci = 0, cl = contiguous.length;
+        int z = itemZ[0];
+        int startI = 0;
+        // Point3: <z, index, length>
+        for(int i = 1; i < numRequests; i++){
+            if(itemZ[i] != z){ // if contiguous section should end
+                contiguous[ci] = z;
+                contiguous[ci + 1] = startI;
+                contiguous[ci + 2] = i - startI;
+                ci += 3;
+                if(ci + 3 > cl){
+                    contiguous = Arrays.copyOf(contiguous, cl <<= 1);
+                }
+                z = itemZ[startI = i];
+            }
+        }
+        contiguous[ci] = z;
+        contiguous[ci + 1] = startI;
+        contiguous[ci + 2] = numRequests - startI;
+        this.contiguous = contiguous;
+
+        final int L = (ci / 3) + 1;
+
+        if(contiguousCopy.length < contiguous.length) this.contiguousCopy = new int[contiguous.length];
+
+        final int[] sorted = CountingSort.countingSortMapMT(contiguous, contiguousCopy, L);
+
+
+        final int[] locs = contiguous;
+        locs[0] = 0;
+        for(int i = 0, ptr = 0; i < L; i++){
+            ptr += sorted[i * 3 + 2];
+            locs[i + 1] = ptr;
+        }
+        if(copy.length < requests.length) copy = new DrawRequest[requests.length];
+        PopulateTask.tasks = sorted;
+        PopulateTask.src = requests;
+        PopulateTask.dest = copy;
+        PopulateTask.locs = locs;
+        commonPool.pool.invoke(new PopulateTask(0, L));
+    }
+
+    protected void sortRequestsStandard(){ // Non-threaded implementation for weak devices
+        final int numRequests = this.numRequests;
+        final int[] itemZ = requestZ;
+        int[] contiguous = this.contiguous;
+        int ci = 0, cl = contiguous.length;
+        int z = itemZ[0];
+        int startI = 0;
+        // Point3: <z, index, length>
+        for(int i = 1; i < numRequests; i++){
+            if(itemZ[i] != z){ // if contiguous section should end
+                contiguous[ci] = z;
+                contiguous[ci + 1] = startI;
+                contiguous[ci + 2] = i - startI;
+                ci += 3;
+                if(ci + 3 > cl){
+                    contiguous = Arrays.copyOf(contiguous, cl <<= 1);
+                }
+                z = itemZ[startI = i];
+            }
+        }
+        contiguous[ci] = z;
+        contiguous[ci + 1] = startI;
+        contiguous[ci + 2] = numRequests - startI;
+        this.contiguous = contiguous;
+
+        final int L = (ci / 3) + 1;
+
+        if(contiguousCopy.length < contiguous.length) contiguousCopy = new int[contiguous.length];
+
+        final int[] sorted = CountingSort.countingSortMap(contiguous, contiguousCopy, L);
+
+        if(copy.length < numRequests) copy = new DrawRequest[numRequests + (numRequests >> 3)];
+        int ptr = 0;
+        final DrawRequest[] items = requests, dest = copy;
+        for(int i = 0; i < L * 3; i += 3){
+            final int pos = sorted[i + 1], length = sorted[i + 2];
+            if(length < 10){
+                final int end = pos + length;
+                for(int sj = pos, dj = ptr; sj < end; sj++, dj++){
+                    dest[dj] = items[sj];
+                }
+            }else System.arraycopy(items, pos, dest, ptr, Math.min(length, dest.length - ptr));
+            ptr += length;
+        }
+    }
+
+    static class CountingSort{
+        private static final int processors = Runtime.getRuntime().availableProcessors() * 8;
+
+        static int[] locs = new int[100];
+        static final int[][] locses = new int[processors][100];
+
+        static final IntIntMap[] countses = new IntIntMap[processors];
+
+        private static Point2[] entries = new Point2[100];
+
+        private static int[] entries3 = new int[300], entries3a = new int[300];
+        private static Integer[] entriesBacking = new Integer[100];
+
+        private static final CountingSortTask[] tasks = new CountingSortTask[processors];
+        private static final CountingSortTask2[] task2s = new CountingSortTask2[processors];
+        private static final Future<?>[] futures = new Future<?>[processors];
+
+        static{
+            for(int i = 0; i < countses.length; i++) countses[i] = new IntIntMap();
+            for(int i = 0; i < entries.length; i++) entries[i] = new Point2();
+
+            for(int i = 0; i < processors; i++){
+                tasks[i] = new CountingSortTask();
+                task2s[i] = new CountingSortTask2();
+            }
+        }
+
+        static class CountingSortTask implements Runnable{
+            static int[] arr;
+            int start, end, id;
+
+            public void set(int start, int end, int id){
+                this.start = start;
+                this.end = end;
+                this.id = id;
+            }
+
+            @Override
+            public void run(){
+                final int id = this.id, start = this.start, end = this.end;
+                int[] locs = locses[id];
+                final int[] arr = CountingSortTask.arr;
+                final IntIntMap counts = countses[id];
+                counts.clear();
+                int unique = 0;
+                for(int i = start; i < end; i++){
+                    int loc = counts.getOrPut(arr[i * 3], unique);
+                    arr[i * 3] = loc;
+                    if(loc == unique){
+                        if(unique >= locs.length){
+                            locs = Arrays.copyOf(locs, unique * 3 / 2);
+                        }
+                        locs[unique++] = 1;
+                    }else{
+                        locs[loc]++;
+                    }
+                }
+                locses[id] = locs;
+            }
+        }
+
+        static class CountingSortTask2 implements Runnable{
+            static int[] src, dest;
+            int start, end, id;
+
+            public void set(int start, int end, int id){
+                this.start = start;
+                this.end = end;
+                this.id = id;
+            }
+
+            @Override
+            public void run(){
+                final int start = this.start, end = this.end;
+                final int[] locs = locses[id];
+                final int[] src = CountingSortTask2.src, dest = CountingSortTask2.dest;
+                for(int i = end - 1, i3 = i * 3; i >= start; i--, i3 -= 3){
+                    final int destPos = --locs[src[i3]] * 3;
+                    dest[destPos] = src[i3];
+                    dest[destPos + 1] = src[i3 + 1];
+                    dest[destPos + 2] = src[i3 + 2];
+                }
+            }
+        }
+
+        static int[] countingSortMapMT(final int[] arr, final int[] swap, final int end){
+            final IntIntMap[] countses = CountingSort.countses;
+            final int[][] locs = CountingSort.locses;
+            final int threads = Math.min(processors, (end + 4095) / 4096); // 4096 Point3s to process per thread
+            final int thread_size = end / threads + 1;
+            final CountingSortTask[] tasks = CountingSort.tasks;
+            final CountingSortTask2[] task2s = CountingSort.task2s;
+            final Future<?>[] futures = CountingSort.futures;
+            CountingSortTask.arr = CountingSortTask2.src = arr;
+            CountingSortTask2.dest = swap;
+
+            for(int s = 0, thread = 0; thread < threads; thread++, s += thread_size){
+                CountingSortTask task = tasks[thread];
+                final int stop = Math.min(s + thread_size, end);
+                task.set(s, stop, thread);
+                task2s[thread].set(s, stop, thread);
+                futures[thread] = commonPool.pool.submit(task);
+            }
+
+            int unique = 0;
+            for(int i = 0; i < threads; i++){
+                try{
+                    futures[i].get();
+                }catch(ExecutionException | InterruptedException e){
+                    commonPool.pool.execute(tasks[i]);
+                }
+                unique += countses[i].size;
+            }
+
+            final int L = unique;
+            if(entriesBacking.length < L){
+                entriesBacking = new Integer[L * 3 / 2];
+                entries3 = new int[L * 3 * 3 / 2];
+                entries3a = new int[L * 3 * 3 / 2];
+            }
+            final int[] entries = CountingSort.entries3, entries3a = CountingSort.entries3a;
+            final Integer[] entriesBacking = CountingSort.entriesBacking;
+            int j = 0;
+            for(int i = 0; i < threads; i++){
+                if(countses[i].size == 0) continue;
+                final IntIntMap.Entries countEntries = countses[i].entries();
+                final IntIntMap.Entry entry = countEntries.next();
+                entries[j] = entry.key;
+                entries[j + 1] = entry.value;
+                entries[j + 2] = i;
+                j += 3;
+                while(countEntries.hasNext){
+                    countEntries.next();
+                    entries[j] = entry.key;
+                    entries[j + 1] = entry.value;
+                    entries[j + 2] = i;
+                    j += 3;
+                }
+            }
+
+            for(int i = 0; i < L; i++){
+                entriesBacking[i] = i;
+            }
+            Arrays.sort(entriesBacking, 0, L, Structs.comparingInt(i -> entries[i * 3]));
+            for(int i = 0; i < L; i++){
+                int from = entriesBacking[i] * 3, to = i * 3;
+                entries3a[to] = entries[from];
+                entries3a[to + 1] = entries[from + 1];
+                entries3a[to + 2] = entries[from + 2];
+            }
+
+            for(int i = 0, pos = 0; i < L * 3; i += 3){
+                pos = (locs[entries3a[i + 2]][entries3a[i + 1]] += pos);
+            }
+
+            for(int thread = 0; thread < threads; thread++){
+                futures[thread] = commonPool.pool.submit(task2s[thread]);
+            }
+            for(int i = 0; i < threads; i++){
+                try{
+                    futures[i].get();
+                }catch(ExecutionException | InterruptedException e){
+                    commonPool.pool.execute(task2s[i]);
+                }
+            }
+            return swap;
+        }
+
+        static int[] countingSortMap(final int[] arr, final int[] swap, final int end){
+            int[] locs = CountingSort.locs;
+            final IntIntMap counts = CountingSort.countses[0];
+            counts.clear();
+
+            int unique = 0;
+            final int end3 = end * 3;
+            for(int i = 0; i < end3; i += 3){
+                int loc = counts.getOrPut(arr[i], unique);
+                arr[i] = loc;
+                if(loc == unique){
+                    if(unique >= locs.length){
+                        locs = Arrays.copyOf(locs, unique * 3 / 2);
+                    }
+                    locs[unique++] = 1;
+                }else{
+                    locs[loc]++;
+                }
+            }
+            CountingSort.locs = locs;
+
+            if(entries.length < unique){
+                final int prevLength = entries.length;
+                entries = Arrays.copyOf(entries, unique * 3 / 2);
+                final Point2[] entries = CountingSort.entries;
+                for(int i = prevLength; i < entries.length; i++) entries[i] = new Point2();
+            }
+            final Point2[] entries = CountingSort.entries;
+
+            final IntIntMap.Entries countEntries = counts.entries();
+            final IntIntMap.Entry entry = countEntries.next();
+            entries[0].set(entry.key, entry.value);
+            int j = 1;
+            while(countEntries.hasNext){
+                countEntries.next(); // it returns the same entry over and over again.
+                entries[j++].set(entry.key, entry.value);
+            }
+            Arrays.sort(entries, 0, unique, Structs.comparingInt(p -> p.x));
+
+            int prev = entries[0].y, next;
+            for(int i = 1; i < unique; i++){
+                locs[next = entries[i].y] += locs[prev];
+                prev = next;
+            }
+            for(int i = end - 1, i3 = i * 3; i >= 0; i--, i3 -= 3){
+                final int destPos = --locs[arr[i3]] * 3;
+                swap[destPos] = arr[i3];
+                swap[destPos + 1] = arr[i3 + 1];
+                swap[destPos + 2] = arr[i3 + 2];
+            }
+            return swap;
+        }
+    }
+
+    static class PopulateTask extends RecursiveAction{
+        int from, to;
+        static int[] tasks;
+        static DrawRequest[] src;
+        static DrawRequest[] dest;
+        static int[] locs;
+
+        //private static final int threshold = 256;
+        PopulateTask(int from, int to){
+            this.from = from;
+            this.to = to;
+        }
+
+        @Override
+        protected void compute(){
+            final int[] locs = PopulateTask.locs;
+            if(to - from > 1 && locs[to] - locs[from] > 2048){
+                final int half = (locs[to] + locs[from]) >> 1;
+                int mid = Arrays.binarySearch(locs, from, to, half);
+                if(mid < 0) mid = -mid - 1;
+                if(mid != from && mid != to){
+                    invokeAll(new PopulateTask(from, mid), new PopulateTask(mid, to));
+                    return;
+                }
+            }
+            final DrawRequest[] src = PopulateTask.src, dest = PopulateTask.dest;
+            final int[] tasks = PopulateTask.tasks;
+            for(int i = from; i < to; i++){
+                final int point = i * 3, pos = tasks[point + 1], length = tasks[point + 2];
+                if(length < 10){
+                    final int end = pos + length;
+                    for(int sj = pos, dj = locs[i]; sj < end; sj++, dj++){
+                        dest[dj] = src[sj];
+                    }
+                }else{
+                    System.arraycopy(src, pos, dest, locs[i], Math.min(length, dest.length - locs[i]));
+                }
+            }
+        }
+    }
+}
diff --git a/core/src/mindustry/graphics/MultiPacker.java b/core/src/mindustry/graphics/MultiPacker.java
index e5f473a7651495f7f61856e6df711fba12dfb1be..f1a3f1cbbac0fbda6dfabe689f25c2108ef58337 100644
--- a/core/src/mindustry/graphics/MultiPacker.java
+++ b/core/src/mindustry/graphics/MultiPacker.java
@@ -114,7 +114,7 @@ public class MultiPacker implements Disposable{
     //editor page (sprites5.png) - all sprites needed for rendering in the editor, including block icons and a few minor sprites
     public enum PageType{
         //main page can be massive, but 8192 throws GL_OUT_OF_MEMORY on some GPUs and I can't deal with it yet.
-        main(4096),
+        main(16384, 4096), //MDTX 更大图集
 
         //TODO stuff like this throws OOM on some devices
         environment(4096, 2048),
diff --git a/core/src/mindustryX/features/RenderExt.java b/core/src/mindustryX/features/RenderExt.java
index 7b5d21f77cd8d15d0541cd70a051aad660c21e1d..e3721adf4c7e215ef3b944512d4608922d22fd11 100644
--- a/core/src/mindustryX/features/RenderExt.java
+++ b/core/src/mindustryX/features/RenderExt.java
@@ -30,6 +30,7 @@ public class RenderExt{
     public static float overdriveZoneTransparency, mendZoneTransparency;
     public static boolean logicDisplayNoBorder, arcDrillMode;
     public static int blockRenderLevel;
+    public static boolean renderSort, renderMerge;
 
     public static boolean unitHide = false;
     private static Effect placementEffect;
@@ -60,6 +61,8 @@ public class RenderExt{
             logicDisplayNoBorder = Core.settings.getBool("arclogicbordershow");
             arcDrillMode = Core.settings.getBool("arcdrillmode");
             blockRenderLevel = Core.settings.getInt("blockRenderLevel");
+            renderSort = Core.settings.getBool("renderSort");
+            renderMerge = Core.settings.getBool("renderMerge");
         });
         Events.run(Trigger.draw, RenderExt::draw);
         Events.on(TileChangeEvent.class, RenderExt::onSetBlock);
